sw/device/tock/dif/build.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+extern crate bindgen;
+
+use bindgen::builder;
+use std::env;
+use std::fs::{self};
+use std::path::Path;
+
+fn main() {
+    let source_root = env::var("MESON_SOURCE_ROOT").unwrap();
+    let build_root = env::var("MESON_BUILD_ROOT").unwrap();
+    let out_dir = env::var("OUT_DIR").unwrap();
+
+    generate_bindings(&source_root, &build_root, &out_dir);
+
+    let dif_lib_path = Path::new(&build_root).join("sw/device/lib/dif");
+    let dif_libs = &["uart_ot"];
+    libraries_link(dif_lib_path.to_str().unwrap(), dif_libs);
+
+    let mmio_lib_path = Path::new(&build_root).join("sw/device/lib/base");
+    let base_libs = &["mmio_ot"];
+    libraries_link(mmio_lib_path.to_str().unwrap(), base_libs);
+}
+
+fn libraries_link(path: &str, libs: &[&str]) -> () {
+    for lib in libs {
+        println!("cargo:rustc-link-search={}", path);
+        println!("cargo:rustc-link-lib={}", lib);
+    }
+}
+
+fn generate_bindings(source_root: &String, build_root: &String, out_dir: &String) {
```

Comment:
Since you use these exclusively as paths, consider accepting `impl AsRef<Path>` instead; you can then do `source_root.as_ref().join("blah");` below.

Reply:
It then complains about `.clang_args(&["-I", &source_root])`, so I kept arguments as `&str`.